File: frontend/src/lib/api/git.ts
/**
 * Git and Git LFS API client
 */

import { api } from './client';

export interface GitCommit {
  hash: string;
  message: string;
  author: string;
  email: string;
  date: string;
  files_changed: number;
}

export interface GitStatus {
  initialized: boolean;
  branch?: string;
  is_dirty?: boolean;
  untracked_files?: string[];
  modified_files?: string[];
  staged_files?: string[];
  lfs_files?: string[];
  lfs_patterns?: string[];
  error?: string;
}

export interface LFSFile {
  oid: string;
  size: number;
  path: string;
}

export interface LFSStatus {
  enabled: boolean;
  installed: boolean;
  initialized?: boolean;
  tracked_patterns?: string[];
  tracked_files?: LFSFile[];
  file_count?: number;
  total_size?: number;
  error?: string;
}

export interface LFSValidation {
  git_installed: boolean;
  lfs_installed: boolean;
  git_version?: string;
  lfs_version?: string;
  issues: string[];
}

export interface CommitRequest {
  message: string;
  prefix?: string;
  files?: string[];
}

export interface EnhancedGitCommit extends GitCommit {
  hash: string;
  shortHash: string;
  stats: {
    additions: number;
    deletions: number;
    files: number;
  };
  files: Array<{
    path: string;
    changeType: string;
    additions: number;
    deletions: number;
  }>;
  parentHashes: string[];
}

export interface RollbackRequest {
  commitHash: string;
  mode: 'soft' | 'mixed' | 'hard';
}

export interface TagRequest {
  tagName: string;
  message?: string;
}

export const gitApi = {
  async getStatus(projectId: string): Promise<GitStatus> {
    return api.get<GitStatus>(`/git/projects/${projectId}/status`);
  },

  async commit(projectId: string, request: CommitRequest): Promise<void> {
    await api.post(`/git/projects/${projectId}/commit`, request);
  },

  async getHistory(projectId: string, limit = 20, file?: string): Promise<GitCommit[]> {
    const params = new URLSearchParams({ limit: limit.toString() });
    if (file) params.append('file', file);
    return api.get<GitCommit[]>(`/git/projects/${projectId}/history?${params}`);
  },

  async getEnhancedHistory(
    projectId: string,
    limit = 50,
    file?: string
  ): Promise<EnhancedGitCommit[]> {
    const params = new URLSearchParams({ limit: limit.toString() });
    if (file) params.append('file_path', file);
    return api.get<EnhancedGitCommit[]>(`/git/projects/${projectId}/history/enhanced?${params}`);
  },

  async rollback(projectId: string, request: RollbackRequest): Promise<void> {
    await api.post(`/git/projects/${projectId}/rollback`, request);
  },

  async createTag(projectId: string, request: TagRequest): Promise<void> {
    await api.post(`/git/projects/${projectId}/tags`, request);
  },

  async trackFileChange(projectId: string, filePath: string): Promise<void> {
    await api.post(`/git/projects/${projectId}/auto-commit/${encodeURIComponent(filePath)}`);
  },

  async forceAutoCommitAll(): Promise<void> {
    await api.post('/git/auto-commit/force-all');
  }
};

export const gitLFSApi = {
  async validateSetup(): Promise<LFSValidation> {
    return api.get<LFSValidation>('/git/lfs/validate');
  },

  async initializeProject(projectId: string): Promise<void> {
    await api.post(`/git/lfs/projects/${projectId}/initialize`);
  },

  async getStatus(projectId: string): Promise<LFSStatus> {
    return api.get<LFSStatus>(`/git/lfs/projects/${projectId}/status`);
  },

  async trackPattern(projectId: string, pattern: string): Promise<void> {
    await api.post(`/git/lfs/projects/${projectId}/track`, { pattern });
  },

  async untrackPattern(projectId: string, pattern: string): Promise<void> {
    await api.post(`/git/lfs/projects/${projectId}/untrack`, { pattern });
  },

  async getTrackedFiles(projectId: string): Promise<LFSFile[]> {
    return api.get<LFSFile[]>(`/git/lfs/projects/${projectId}/files`);
  }
};

File: backend/app/schemas/git.py
"""
Pydantic schemas for Git and Git LFS operations.
"""

from pydantic import BaseModel, Field


class GitCommit(BaseModel):
    """Git commit information"""

    hash: str = Field(..., description="Commit hash (short)")
    message: str = Field(..., description="Commit message")
    author: str = Field(..., description="Author name")
    email: str = Field(..., description="Author email")
    date: str = Field(..., description="Commit date in ISO format")
    files_changed: int = Field(..., description="Number of files changed")


class GitStatus(BaseModel):
    """Git repository status"""

    initialized: bool = Field(..., description="Whether repository is initialized")
    branch: str | None = Field(None, description="Current branch name")
    is_dirty: bool | None = Field(None, description="Whether repository has uncommitted changes")
    untracked_files: list[str] | None = Field(None, description="List of untracked files")
    modified_files: list[str] | None = Field(None, description="List of modified files")
    staged_files: list[str] | None = Field(None, description="List of staged files")
    lfs_files: list[str] | None = Field(None, description="List of LFS-tracked files")
    lfs_patterns: list[str] | None = Field(None, description="List of LFS tracking patterns")
    error: str | None = Field(None, description="Error message if any")


class LFSFile(BaseModel):
    """Git LFS tracked file information"""

    oid: str = Field(..., description="Object ID")
    size: int = Field(..., description="File size in bytes")
    path: str = Field(..., description="File path")


class LFSStatus(BaseModel):
    """Git LFS status for a project"""

    enabled: bool = Field(..., description="Whether LFS is enabled")
    installed: bool = Field(..., description="Whether LFS is installed")
    initialized: bool | None = Field(None, description="Whether LFS is initialized in project")
    tracked_patterns: list[str] | None = Field(None, description="Patterns tracked by LFS")
    tracked_files: list[LFSFile] | None = Field(None, description="Files tracked by LFS")
    file_count: int | None = Field(None, description="Number of LFS-tracked files")
    total_size: int | None = Field(None, description="Total size of LFS-tracked files")
    error: str | None = Field(None, description="Error message if any")


class LFSTrackRequest(BaseModel):
    """Request to track a file pattern with Git LFS"""

    pattern: str = Field(..., description="File pattern to track (e.g., *.mp4)")


class LFSValidation(BaseModel):
    """Git LFS installation validation result"""

    git_installed: bool = Field(..., description="Whether Git is installed")
    lfs_installed: bool = Field(..., description="Whether Git LFS is installed")
    git_version: str | None = Field(None, description="Git version string")
    lfs_version: str | None = Field(None, description="Git LFS version string")
    issues: list[str] = Field(default_factory=list, description="List of issues found")


class CommitRequest(BaseModel):
    """Request to commit changes"""

    message: str = Field(..., description="Commit message")
    prefix: str | None = Field(None, description="Semantic commit prefix (feat, fix, etc.)")
    files: list[str] | None = Field(None, description="Specific files to commit")


class RepositoryValidation(BaseModel):
    """Repository validation result"""

    valid: bool = Field(..., description="Whether repository is valid")
    issues: list[str] = Field(default_factory=list, description="Critical issues found")
    warnings: list[str] = Field(default_factory=list, description="Warnings found")
